compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -200,10 +200,24 @@ fn layout_of_uncached<'tcx>(
                     return Ok(tcx.mk_layout(LayoutS::scalar(cx, data_ptr)));
                 }
 
-                let Abi::Scalar(metadata) = metadata_layout.abi else {
+                let Abi::Scalar(mut metadata) = metadata_layout.abi else {
                     return Err(error(cx, LayoutError::Unknown(pointee)));
                 };
 
+                if !ty.is_unsafe_ptr() {
+                    match pointee.kind() {
+                        ty::Slice(element) => {
```

Comment:
So `&[i32]` is getting the optimization but `&(bool, [i32])` is not? That seems odd?